libdnf/rpm/package_set.cpp

```diff
@@ -0,0 +1,255 @@
+    map_init_clone(&map, &pset.pImpl->map);
+}
+
+
+PackageSet::Impl::~Impl() {
+    map_free(&map);
+}
+
+
+// see http://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetTable
+static const unsigned char _BitCountLookup[256] = {
+    #define B2(n) n, n + 1, n + 1, n + 2
+    #define B4(n) B2(n), B2(n + 1), B2(n + 1), B2(n + 2)
+    #define B6(n) B4(n), B4(n + 1), B4(n + 1), B4(n + 2)
+    B6(0), B6(1), B6(1), B6(2)
+};
+
+
+inline size_t map_count(Map * m) {
+    unsigned char * ti = m->map;
+    unsigned char * end = ti + m->size;
+    unsigned c = 0;
+
+    while (ti < end) {
+        c += _BitCountLookup[*ti++];
+    }
+
+    return c;
+}
+
+
+Id PackageSet::operator[](unsigned int index) const {
+    const unsigned char * ti = pImpl->map.map;
+    const unsigned char * end = ti + pImpl->map.size;
+    unsigned int enabled;
+    Id id;
+
+    while (ti < end) {
+        enabled = _BitCountLookup[*ti];
+
+        if (index >= enabled) {
+            index -= enabled;
+            ti++;
+            continue;
+        }
+        id = static_cast<Id>(ti - pImpl->map.map) << 3;
+
+        index++;
+        for (unsigned char byte = *ti; index; byte >>= 1) {
+            if ((byte & 0x01))
+                index--;
+            if (index)
+                id++;
+        }
+        return id;
+    }
+    return -1;
+}
+
+
+PackageSet & PackageSet::operator|=(const PackageSet & other) {
+    map_or(&pImpl->map, &other.pImpl->map);
+    return *this;
+}
+
+
+PackageSet & PackageSet::operator-=(const PackageSet & other) {
+    map_subtract(&pImpl->map, &other.pImpl->map);
+    return *this;
+}
+
+
+PackageSet & PackageSet::operator&=(const PackageSet & other) {
+    map_and(&pImpl->map, &other.pImpl->map);
+    return *this;
+}
+
+
+PackageSet & PackageSet::operator|=(const Map * other) {
+    map_or(&pImpl->map, const_cast<Map *>(other));
+    return *this;
+}
+
+PackageSet & PackageSet::operator-=(const Map * other) {
+    map_subtract(&pImpl->map, const_cast<Map *>(other));
+    return *this;
+}
+
+
+PackageSet & PackageSet::operator&=(const Map * other) {
+    map_and(&pImpl->map, const_cast<Map *>(other));
+    return *this;
+}
+
+
+void PackageSet::clear() {
+    map_empty(&pImpl->map);
+}
+
+
+bool PackageSet::empty() {
+    const unsigned char * res = pImpl->map.map;
+    const unsigned char * end = res + pImpl->map.size;
+
+    while (res < end) {
+        if (*res++)
+            return false;
+    }
+    return true;
+}
+
+
+void PackageSet::set(const Package & pkg) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wconversion"
+    MAPSET(&pImpl->map, pkg.get_id());
+#pragma GCC diagnostic pop
+}
+
+
+void PackageSet::set(Id id) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wconversion"
+    MAPSET(&pImpl->map, id);
+#pragma GCC diagnostic pop
+}
+
+
+bool PackageSet::has(const Package & pkg) const {
+    return MAPTST(&pImpl->map, pkg.get_id());
```

Comment:
We need two additional public methods that will check range. You should not set a bit outside of range of map size. This method could be only private when we will do a test for multiple entries.